zebra-test/src/mock_service.rs

```diff
@@ -0,0 +1,731 @@
+/// amount of time, so this may affect the test execution time.
+const DEFAULT_MAX_REQUEST_DELAY: Duration = Duration::from_millis(25);
+
+/// An internal type representing the item that's sent in the [`broadcast`] channel.
+///
+/// The actual type that matters is the [`ResponseSender`] but since there could be more than one
+/// [`MockService`] verifying requests, the type must be wrapped so that it can be shared by all
+/// receivers:
+///
+/// - The [`Arc`] makes sure the instance is on the heap, and can be shared properly between
+///   threads and dropped when no longer needed.
+/// - The [`Mutex`] ensures only one [`MockService`] instance can reply to the received request.
+/// - The [`Option`] forces the [`MockService`] that handles the request to take ownership of it
+///   because sending a response also forces the [`ResponseSender`] to be dropped.
+type ProxyItem<Request, Response, Error> =
+    Arc<Mutex<Option<ResponseSender<Request, Response, Error>>>>;
+
+/// A service implementation that allows intercepting requests for checking them.
+///
+/// The type is generic over the request and response types, and also has an extra generic type
+/// parameter that's used as a tag to determine if the internal assertions should panic or return
+/// errors for proptest minimization. See [`AssertionType`] for more information.
+///
+/// The mock service can be cloned, and provides methods for checking the received requests as well
+/// as responding to them individually.
+///
+/// Internally, the instance that's operating as the service will forward requests to a
+/// [`broadcast`] channel that the other instances listen to.
+///
+/// See the [module-level documentation][`super::mock_service`] for an example.
+pub struct MockService<Request, Response, Assertion, Error = BoxError> {
+    receiver: broadcast::Receiver<ProxyItem<Request, Response, Error>>,
+    sender: broadcast::Sender<ProxyItem<Request, Response, Error>>,
+    max_request_delay: Duration,
+    _assertion_type: PhantomData<Assertion>,
+}
+
+/// A builder type to create a [`MockService`].
+///
+/// Allows changing specific parameters used by the [`MockService`], if necessary. The default
+/// parameters should be reasonable for most cases.
+#[derive(Default)]
+pub struct MockServiceBuilder {
+    proxy_channel_size: Option<usize>,
+    max_request_delay: Option<Duration>,
+}
+
+/// A helper type for responding to incoming requests.
+///
+/// An instance of this type is created for each request received by the [`MockService`]. It
+/// contains the received request and a [`oneshot::Sender`] that can be used to respond to the
+/// request.
+///
+/// If a response is not sent, the channel is closed and a [`BoxError`] is returned by the service
+/// to the caller that sent the request.
+#[must_use = "Tests may fail if a response is not sent back to the caller"]
+pub struct ResponseSender<Request, Response, Error> {
+    request: Request,
```

Comment:
In some cases it may necessary to access the request to generate the response, so maybe make this `pub` or create an getter for it

Reply:
Done in [dd608b0](https://github.com/ZcashFoundation/zebra/pull/2748/commits/dd608b0113e5ae2e01dd721f89e222346a77c40f).